src/nvim/extmark.c

```diff
@@ -194,25 +194,32 @@ static bool extmark_setraw(buf_T *buf, uint64_t mark, int row, colnr_T col)
   return true;
 }
 
-/// Remove an extmark
+/// Remove an extmark pointed to by "itr" or from "ns_id" and "id"
 ///
-/// @return  0 on missing id
-bool extmark_del(buf_T *buf, uint32_t ns_id, uint32_t id)
+/// @return  line number of the deleted mark, -1 on missing id
+linenr_T extmark_del(buf_T *buf, MarkTreeIter *itr, MTKey key, uint32_t ns_id, uint32_t id)
 {
-  MarkTreeIter itr[1] = { 0 };
-  MTKey key = marktree_lookup_ns(buf->b_marktree, ns_id, id, false, itr);
-  if (!key.id) {
-    return false;
+  MarkTreeIter it[1] = { 0 };
+
+  if (itr == NULL) {
+    key = marktree_lookup_ns(buf->b_marktree, ns_id, id, false, it);
+    if (!key.id) {
+      return 0;
+    }
+    assert(key.pos.row >= 0);
+    itr = it;
```

Comment:
I would suggest breaking this part out to a separate wrapper function, so we have `marktree_del(buf_T *buf, MarkTreeIter *itr, MTKey key, uint32_t ns_id, uint32_t id)` and `marktree_del_itr((buf_T *buf, MarkTreeIter *itr, MTKey key)` call sites which use all of their arguments.

Reply:
I ended up with different function signatures when breaking it out but done😅 This what you had in mind?